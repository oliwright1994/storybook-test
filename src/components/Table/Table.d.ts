import { Column } from 'react-table'

export type TData = object

export interface ITable {
  columns: Column<Data>[]
  data: Data[]
}
