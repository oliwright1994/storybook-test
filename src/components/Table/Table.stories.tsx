import { Button } from '@chakra-ui/core';
import React from 'react';
import { CellProps } from 'react-table';

import createStory from '../../utils/createStory';
import Tag from '../Tag';
import Table from '.';

createStory(
  'COMPONENTS | Table',
  [
    {
      label: 'inbox example',
      jsx: () => (
        <Table
          columns={[
            {
              Header: 'Date',
              accessor: 'date',
            },
            {
              Header: 'From',
              accessor: 'sender',
            },
            {
              Header: '',
              accessor: 'tags',
              Cell: ({ row }: CellProps<object>) =>
                row.values.tags.map((tag: any) => (
                  <span style={{ marginRight: 2 }}>
                    <Tag messageType={tag.type} key={tag.label}>
                      {tag.label}
                    </Tag>
                  </span>
                )),
            },
            {
              Header: 'Subject',
              accessor: 'subject',
            },
          ]}
          data={[
            {
              date: '14:20',
              sender: 'Smart',
              tags: [{ type: 'info', label: 'Unread' }],
              subject: "This is the subject of the message you've received",
            },
            {
              date: '15:20',
              sender: 'Smart',
              tags: [
                { type: 'info', label: 'Unread' },
                { type: 'warning', label: 'Important' },
              ],
              subject: "You\'ve got 1 month's income left",
            },
            {
              date: '16:20',
              sender: 'Smart',
              tags: [],
              subject: "Your income is due to run out when you're 76. Here is what to do if you want your blah ",
            },
            {
              date: '17:20',
              sender: 'Smart',
              tags: [],
              subject: 'Mauris henderit suspicit ante',
            },
            {
              date: '18:20',
              sender: 'Smart',
              tags: [{ type: 'warning', label: 'Important' }],
              subject: 'Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus',
            },
          ]}
        />
      ),
    },
    {
      label: 'transaction history example',
      jsx: () => (
        <Table
          columns={[
            { Header: 'Date', accessor: 'date' },
            {
              Header: 'From',
              accessor: 'from',
            },
            {
              Header: 'To',
              accessor: 'to',
            },
            {
              Header: 'Amount',
              accessor: 'amount',
            },
            {
              Header: '',
              accessor: 'actions',
              Cell: ({ row }: CellProps<object>) =>
                row.values.actions.map((action: string) => <Button height="auto">{action}</Button>),
            },
          ]}
          data={[
            {
              date: '23 Sep 2019',
              from: 'Rainy day pot',
              to: 'Flexible income pot',
              amount: '£2,000.00',
              actions: ['Cancel transaction'],
            },
            {
              date: '23 Sep 2019',
              from: 'Inheritance pot',
              to: 'Later life pot',
              amount: '£2,000.00',
              actions: ['Cancel transaction'],
            },
            {
              date: '23 Sep 2019',
              from: 'Flexible income potpot',
              to: 'Paid out',
              amount: '-£225.00',
              actions: ['Cancel transaction'],
            },
          ]}
        />
      ),
    },
  ],
  {}
);
