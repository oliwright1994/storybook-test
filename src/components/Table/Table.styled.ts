import styled from 'styled-components'
import {
  border,
  BorderProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps,
  layout,
  LayoutProps,
  padding,
  PaddingProps,
  typography,
  TypographyProps,
} from 'styled-system'

export const STableWrap = styled.table`
  border-collapse: collapse;
`

export const STableHead = styled.thead``
export const STableBody = styled.tbody``
export const STableHeadCell = styled.th<TypographyProps & PaddingProps & FontWeightProps & FontSizeProps>`
  ${typography}
  ${padding}
  ${fontWeight}
  ${fontSize}
`

export const STableRow = styled.tr<BorderProps & PaddingProps & LayoutProps>`
  ${border}
  ${padding}
  ${layout}
`

export const STableCell = styled.td<LayoutProps & PaddingProps & FontSizeProps>`
  ${layout}
  ${padding}
  ${fontSize}
`
