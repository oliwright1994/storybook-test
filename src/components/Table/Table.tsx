import React, { FC, useEffect, useRef } from 'react';
import {
  ColumnInstance,
  TableInstance,
  useBlockLayout,
  useResizeColumns,
  UseResizeColumnsColumnProps,
  useTable,
  UseTableCellProps,
} from 'react-table';

import { ITable, TData } from './Table.d';
import {
  Resizer, TableBody, TableCell, TableHead, TableRow, TableWrap,
} from './Table.styled';

interface TableColumn<D extends object = {}>
  extends ColumnInstance<D>,
  UseResizeColumnsColumnProps<D>,
  UseTableCellProps<D> {}

const useStopPropagationOnClick = <T extends any>(reference: T) => {
  useEffect(() => {
    if (reference.current !== null) {
      reference.current.addEventListener('click', (e: MouseEvent) => e.stopPropagation());
    }
  }, []);
};

const ResizerComponent: FC = (properties) => {
  const reference = useRef<HTMLDivElement>(null);
  useStopPropagationOnClick(reference);
  return <Resizer {...properties} ref={reference} />;
};

const defaultColumn = {
  minWidth: 20,
  width: 200,
  maxWidth: 500,
};

const Table: FC<ITable> = ({ columns, data }) => {
  const {
    getTableProps, getTableBodyProps, headerGroups, rows, prepareRow,
  } = useTable<TData>(
    {
      columns,
      defaultColumn,
      data,
    },
    useBlockLayout,
    useResizeColumns,
  );

  return (
    <TableWrap {...getTableProps()} display="inline-block">
      <TableHead>
        {headerGroups.map((headerGroup) => (
          <TableRow
            {...headerGroup.getHeaderGroupProps()}
            borderBottom="2px"
            borderColor="background.1"
            fontWeight="bold"
            px={1}
            py={2}
          >
            {headerGroup.headers.map((c) => {
              const column = (c as unknown) as TableColumn<TData>;
              return (
                <TableCell {...column.getHeaderProps()}>
                  {column.render('Header')}
                  <ResizerComponent {...column.getResizerProps()} />
                </TableCell>
              );
            })}
          </TableRow>
        ))}
      </TableHead>
      <TableBody {...getTableBodyProps()}>
        {rows.map((row) => {
          prepareRow(row);
          return (
            <TableRow {...row.getRowProps()} borderBottom="2px" borderColor="background.1" p={1}>
              {row.cells.map((cell) => <TableCell {...cell.getCellProps()}>{cell.render('Cell')}</TableCell>)}
            </TableRow>
          );
        })}
      </TableBody>
    </TableWrap>
  );
};

export default Table;
