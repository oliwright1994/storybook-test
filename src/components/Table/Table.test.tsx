import React from 'react'
import { render } from '@testing-library/react';
import Table from '.'

const props = {
  columns: [
    {
      title: 'Static Column',
      dataIndex: 'col1',
      key: 'col1'
    },
    {
      title: 'Dynamic Render Column',
      dataIndex: 'col2',
      key: 'col2',
      render: ({ col2 }) => <b>{col2}</b>
    }
  ],
  data: [
    {
      col1: 'lorem',
      col2: 'lipsum',
    },
    {
      col1: 'dolor',
      col2: 'sit'
    }
  ]
}

describe('<Table />', () => {
  it('should render as expected', () => {
    const { asFragment } = render(<Table {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
