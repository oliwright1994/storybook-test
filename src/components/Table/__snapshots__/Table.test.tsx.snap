// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render as expected 1`] = `
<DocumentFragment>
  <table
    class="sc-AxjAm ykYLV"
  >
    <thead>
      <tr
        class="sc-AxhCb gtcCAP"
        height="32"
      >
        <th
          class="sc-AxirZ fXGLll"
        >
          Static Column
        </th>
        <th
          class="sc-AxirZ fXGLll"
        >
          Dynamic Render Column
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-AxhCb gHPYCa"
        height="48"
      >
        <td
          class="sc-AxiKw MbEDr"
        >
          lorem
        </td>
        <td
          class="sc-AxiKw MbEDr"
        >
          <b>
            lipsum
          </b>
        </td>
      </tr>
      <tr
        class="sc-AxhCb gHPYCa"
        height="48"
      >
        <td
          class="sc-AxiKw MbEDr"
        >
          dolor
        </td>
        <td
          class="sc-AxiKw MbEDr"
        >
          <b>
            sit
          </b>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
